src/runtime/metal/metal_allocator.cpp

```diff
@@ -0,0 +1,243 @@
+  
+  allocation_t alloc{ offset, phantom_size };
+  _allocations.insert(alloc);
+  validate_sorted();
+  
+  _used_size += phantom_size;
+  _available_size = _heap->maxAvailableSize(1 << 14);
+  return (void*)(_cpu_base_va + offset);
+}
+
+void metal_heap_block::deallocate(void* usm_pointer) {
+  int64_t offset = (uint64_t)usm_pointer - _cpu_base_va;
+  auto phantom_it = _phantom_buffers.find(offset);
+  assert((phantom_it != _phantom_buffers.end()) && "Pointer did not originate from this heap.");
+  MTL::Buffer* phantom_buffer = (*phantom_it).second.get();
+  int64_t phantom_size = phantom_buffer->allocatedSize();
+  _phantom_buffers.erase(phantom_it);
+  
+  allocation_t expected_alloc{ offset, phantom_size };
+  auto alloc_it = _allocations.lower_bound(expected_alloc);
+  assert((alloc_it != _allocations.end()) && "Allocation not found in sorted list.");
+  assert((*alloc_it == expected_alloc) && "Allocation not found in sorted list.");
+  _allocations.erase(alloc_it);
+  
+  _used_size -= phantom_size;
+  _available_size = _heap->maxAvailableSize(1 << 14);
+}
+
+// TODO: Copy doc comment from Swift prototype that explains why we want to
+// batch pointer searches.
+//
+// TODO: Make a more efficient traversal method after debugging, which is
+// batchable but returns failures. Unless the compiler will always inline
+// this. You need to profile and see how much overhead it invokes.
+//
+// Defer the actual optimization until later. In fact, it can be a follow-up
+// pull request. Just quantify the performance impact.
+int64_t metal_heap_block::get_offset(void* usm_pointer) {
+  int64_t offset = (uint64_t)usm_pointer - _cpu_base_va;
+  allocation_t dummy_alloc{ offset, 0 };
+  auto upper_bound_it = _allocations.upper_bound(dummy_alloc);
+  if (upper_bound_it == _allocations.begin()) {
+    return -1;
+  }
+  allocation_t alloc = *(prev(upper_bound_it));
+  if (alloc.offset <= offset &&
+      offset < alloc.offset + alloc.size) {
+    return offset;
+  } else {
+    return -1;
+  }
+}
+
+void metal_heap_block::validate_sorted() {
+  if (_allocations.size() > 1) {
+    auto start = _allocations.begin();
+    auto end = prev(_allocations.end());
+    for (auto it = start; it != end; ++it) {
+      auto element1 = *it;
+      auto element2 = *(next(it));
+      assert((element1.offset + element1.size <= element2.offset) && "Allocations are not sorted.");
+    }
+  }
+}
+
+int64_t fetch_sysctl_property(const char *name) {
+  int64_t ret = 0;
+  size_t size = sizeof(int64_t);
+  int error = sysctlbyname(name, &ret, &size, NULL, 0);
+  assert((error == 0) && "sysctl failed.");
+  return ret;
+}
+
+int64_t round_up_to_power_of_2(int64_t input) {
+  uint64_t to_count = (0 > (input - 1)) ? 0 : (input - 1);
+  return 1 << (64 - std::__countl_zero(to_count));
+}
+
+int64_t round_down_to_power_of_2(int64_t input) {
+  uint64_t to_count = input;
+  return 1 << (64 - 1 - std::__countl_zero(to_count));
+}
+
+}
```

Comment:
I would like to see some more explanations in this file eventually, but I understand that this is work-in-progress :)

Reply:
I was in the middle of translating my [prototype](https://github.com/philipturner/metal-usm/blob/2f98211939ec7110bf50a755764a5f33f9aa637f/MetalAllocator/MainFile.swift#L302-L979) to C++, and lost interest halfway. This is the most iffy part of the entire backend. @woachk might be better able to review this.

I was having trouble with `std::__countl_zero` though. There's no easy way like `clz` in Metal or `.leadingZeroBitCount` in Swift. What form do you prefer for hipSYCL?

Reply:
You could use `__builtin_clz()` which should be widely available on many compilers (since Metal backend is only going to be built on Apple platforms, it would be sufficient if that builtin is supported by the compilers there. clang has it). C++ 20 introduces `std::countl_zero`.